src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,167 @@
+      positions_list.push_back(RowID{chunk_id, chunk_offset});
+    }
+  }
+}
+
+void TableScan::scan_reference_segment(const std::shared_ptr<ReferenceSegment> segment, PosList& positions_list) {
+  const auto typed_search_value = search_value();
+
+  for (const auto row_id : *segment->pos_list()) {
+    const auto value = segment->get_row_id(row_id);
+    if (variant_is_null(value)) {
+      continue;
+    }
+
+    if (compare(value, typed_search_value)) {
+      positions_list.push_back(row_id);
+    }
+  }
+}
+
+std::shared_ptr<const Table> TableScan::_on_execute() {
+  // get posListPtr, tableptr, columncount, data type of table
+  auto table = _in->get_output();
+  auto column_count = table->column_count();
+  auto data_type = table->column_type(column_id());
+  auto chunk_count = table->chunk_count();
+  auto positions_list = std::make_shared<PosList>();
+
+  Assert(!variant_is_null(search_value()), "Table Scan Error: Can not compare against NULL value");
+
+  // Iterate over all Chunks and decide how to execute the scan depending on the type of segment
+  for (auto chunk_id = ChunkID{0}; chunk_id < chunk_count; chunk_id++) {
+    auto segment = table->get_chunk(chunk_id)->get_segment(column_id());
+
+    resolve_data_type(data_type, [&segment, this, &positions_list, &chunk_id, &table](auto type) {
+      using Type = typename decltype(type)::type;
+
+      const auto typed_value_segment = std::dynamic_pointer_cast<ValueSegment<Type>>(segment);
+      const auto typed_dict_segment = std::dynamic_pointer_cast<DictionarySegment<Type>>(segment);
+      const auto reference_segment = std::dynamic_pointer_cast<ReferenceSegment>(segment);
+
+      Assert(typed_value_segment || typed_dict_segment || reference_segment,
+             "TableScan error: segment is not from value segment, dict segment or reference segment.");
+      if (typed_value_segment) {
+        scan_value_segment(typed_value_segment, *positions_list, chunk_id);
+      } else if (typed_dict_segment) {
+        scan_dict_segment(typed_dict_segment, *positions_list, chunk_id);
+      } else {
+        scan_reference_segment(reference_segment, *positions_list);
+        table = reference_segment->referenced_table();
+      }
+    });
+  }
+
+  // create ref segments and add them
+  auto chunk = std::make_shared<Chunk>();
+  for (auto column_id = ColumnID{0}; column_id < column_count; ++column_id) {
+    auto reference_segment = std::make_shared<ReferenceSegment>(table, column_id, positions_list);
+    chunk->add_segment(reference_segment);
+  }
+
+  auto column_definitions = std::vector<Table::ColumnDefinitionStruct>();
+  for (auto column_id = ColumnID{0}; column_id < column_count; ++column_id) {
+    auto column_definition = Table::ColumnDefinitionStruct{table->column_name(column_id), table->column_type(column_id),
+                                                           table->column_nullable(column_id)};
+    column_definitions.push_back(column_definition);
```

Comment:
The strings in `column_definition` are copied here, you could instead `std::move` the struct